pydoctor/templatewriter/__init__.py

```diff
@@ -5,6 +5,364 @@
+from twisted.web.template import XMLString, Element
+
+from pydoctor.model import System, Documentable
+
+def parse_dom(text: str) -> minidom.Document:
+    """
+    Create a L{minidom} representaton of the XML string. 
+    """
+    try:
+        dom = minidom.parseString(text)
+    except Exception as e:
+        raise ValueError(f"Can't parse XML from text '{text}'. XML documents can have only one root. ") from e
+    else:
+        return dom
+
+class UnsupportedTemplateVersion(Exception):
+    """Raised when custom template is designed for a newer version of pydoctor"""
+    pass
+
+@runtime_checkable
+class IWriter(Protocol):
+    """
+    Interface class for pydoctor output writer. 
+    """
+
+    @overload
+    def __init__(self, htmloutput: str) -> None: ...
+    @overload
+    def __init__(self, htmloutput: str, template_lookup: 'TemplateLookup') -> None: ...
+
+    def prepOutputDirectory(self) -> None:
+        """
+        Called first.
+        """
+        ... 
+
+    def writeModuleIndex(self, system:'System') -> None: 
+        """
+        Called second.
+        """
+        ...
+
+    def writeIndividualFiles(self, obs:Iterable['Documentable']) -> None:
+        """
+        Called last.
+        """
+        ...
+
+
+class Template(abc.ABC):
+    """
+    Represents a pydoctor template file. 
+    
+    It holds references to template information. 
+
+    It's an additionnal level of abstraction to hook to the 
+    rendering system, it stores the loader object that 
+    is going to be reused for each output file using this template. 
+
+    Use L{Template.fromfile} to create Templates. 
+
+    @see: L{TemplateLookup}
+    """
+
+    def __init__(self, path: Path):
+        self.path: Path = path
+        """
+        Template file path
+        """
+        
+        with path.open('r') as f:
+            self.text = f.read()
+            """
+            File text
+            """
+        
+
+    path: Path
+    """
+    File path
+    """
+
+    text: str
+    """
+    File text 
+    """
+    
+    TEMPLATE_FILES_SUFFIX = ['.html', '.css', '.js']
+    
+    @classmethod
+    def fromfile(cls, path: Path) -> Optional['Template']:
+        """
+        Create a concrete template object. 
+        Type depends on the file extension. 
+
+        Warns if the template cannot be created.
+
+        @param path: A L{Path} that should point to a HTML, CSS or JS file. 
+        @returns: The template object or C{None} if file is invalid. 
+        """
+        if not path.is_file():
+            warnings.warn(f"Cannot create Template: {path.as_posix()} is not a file.")
+        elif path.suffix.lower() in cls.TEMPLATE_FILES_SUFFIX:
+
+            if path.suffix.lower() == '.html':
+                return _HtmlTemplate(path=path)
+            else:
+                return _StaticTemplate(path=path)
+        else:
+            warnings.warn(f"Cannot create Template: {path.as_posix()} is not a template file.")
```

Comment:
The users can drop `.docx` files inside their custom template folder, this warning takes care of this situation.

Reply:
Ah, then I didn't understand the mechanism correctly.

So it works like this:
- `TemplateLoader.__init__()` scans the template directory and creates `Template` objects for all files
- `TemplateLoader.get_template()` looks up a name in the dictionary of all templates (2 layers: custom and default)

I thought that `TemplateLoader` would instead fill its dictionary on demand, in which case unsupported file names would never be requested.

I think that the current implementation is actually better, because with on-demand loading the error reporting is less predictable and happens later in the execution. So let's stick with the current approach and keep this warning.

I do think it's useful to mention in the warning message that the file name extension is the reason the file is rejected as a template (as opposed to its contents).

Reply:
The warning will be as follow :
```
warnings.warn(f"Cannot create Template: {path.as_posix()} is not recognized as template file. "
                f"Template files must have one of the following extensions: {', '.join(cls.TEMPLATE_FILES_SUFFIX)}")
```